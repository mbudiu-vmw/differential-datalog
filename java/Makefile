JAR=ddlogapi.jar
# There are other *.class files generated, but
# the remaining ones are for inner classes of these
# so they are always generated together.
CLASS = ddlogapi/DDlogAPI.class \
		ddlogapi/DDlogCommand.class \
		ddlogapi/DDlogRecord.class \
		ddlogapi/DDlogRecCommand.class \
		ddlogapi/DDlogException.class

all: ${JAR}

${JAR}: ${CLASS}
	jar cvf ${JAR} ddlogapi/*.class

%.class: %.java ddlogapi_DDlogAPI.h
	javac $<

ddlogapi_DDlogAPI.h: ddlogapi/DDlogAPI.java
	javac -h . $<

clean:
	rm -f *.class ddlogapi/*.class
