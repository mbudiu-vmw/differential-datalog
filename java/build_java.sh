#!/bin/bash
# Shell script which generates a Java program from a DDlog program and compiles it
# The following environment variables are used

stack install
if command clang -v 2>/dev/null; then
    export CC=clang
else
    export CC=gcc
fi

case $(uname -s) in
    Darwin*)    SHLIBEXT=dylib; JDK_OS=darwin;;
    Linux*)     SHLIBEXT=so; JDK_OS=linux;;
    *)          echo "Unsupported OS"; exit 1
esac

function compile {
    local DLFILE=$1   # basename ddlog program that is being compiled (without .dl suffix)
    local JAVAPROG=$2 # java test program that is being compiled
    local BUILD=$3    # one of "debug" or "release"

    if [ "x${BUILD}" != "xdebug" -a "x${BUILD}" != "xrelease" ]; then
        echo "Third argument of compile must be 'debug' or 'release', not ${BUILD}"
        exit 1
    fi

    local DLPROG=$(basename ${DLFILE} .dl)
    local DLDIR=$(dirname ${DLFILE})
    if [ "x${DLDIR}" == "x" ]; then
        DLDIR="."
    fi

    CLASSPATH=$(pwd)/${DLDIR}/${DLPROG}_ddlog/flatbuf/java:$(pwd)/../ddlogapi.jar:..:$CLASSPATH
    # Compile the span_uuid.dl DDlog program
    ddlog -i ${DLFILE} -L../../lib -j
    # Compile the rust program; generates ../test/datalog_tests/span_ddlog/target/debug/libspan_ddlog.a
    pushd ${DLDIR}/${DLPROG}_ddlog
    if [ "x${BUILD}" == "xrelease" ]; then
        cargo build --features=flatbuf --release
    else
        cargo build --features=flatbuf
    fi
    popd
    # Build the Java library with the DDlog API
    make -C ..

    (cd ${DLDIR}/${DLPROG}_ddlog/flatbuf/java && javac $(ls ddlog/__${DLPROG}/*.java) && javac $(ls ddlog/${DLPROG}/*.java))

    # Compile Java program
    javac -encoding utf8 -Xlint:unchecked ${JAVAPROG}
    # Create a shared library containing all the native code: ddlogapi.c, ${DLPROG}_ddlog.a
    ${CC} -shared -fPIC -I${JAVA_HOME}/include -I${JAVA_HOME}/include/${JDK_OS} -I../../rust/template -I../../lib ../ddlogapi.c -L${DLDIR}/${DLPROG}_ddlog/target/${BUILD}/ -l${DLPROG}_ddlog -o libddlogapi.${SHLIBEXT}
}

function cleanup {
    rm -rf ./*.class libddlogapi.${SHLIBEXT}
}
