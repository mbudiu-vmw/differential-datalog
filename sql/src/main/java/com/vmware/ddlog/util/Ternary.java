package com.vmware.ddlog.util;

public enum Ternary {
    Yes,
    No,
    Maybe
}
