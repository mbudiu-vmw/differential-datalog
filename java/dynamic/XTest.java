import java.util.*;
import ddlogapi.*;

public class XTest {
    private final DDlogAPI api;

    XTest(DDlogAPI api) throws DDlogException {
        this.api = api;
    }

    void onCommit(DDlogCommand command) {
        System.out.println(command.toString());
    }

    void run() throws DDlogException {
        DDlogRecord field = new DDlogRecord(10);
        DDlogRecord[] fields = { field };
        DDlogRecord record = DDlogRecord.makeStruct("R", fields);
        int id = this.api.getTableId("R");
        DDlogRecCommand command = new DDlogRecCommand(
            DDlogCommand.Kind.Insert, id, record);
        DDlogRecCommand[] ca = new DDlogRecCommand[1];
        ca[0] = command;

        System.err.println("Executing " + command.toString());
        this.api.transactionStart();
        this.api.applyUpdates(ca);
        this.api.transactionCommitDumpChanges(this::onCommit);
    }

    public static void main(String[] args) throws DDlogException {
        DDlogAPI api = DDlogAPI.compileAndLoad("x.dl", "../..");
        if (api == null) {
            System.err.println("Could not load DDlogAPI");
            System.exit(1);
        }
        XTest test = new XTest(api);
        test.run();
    }
}
