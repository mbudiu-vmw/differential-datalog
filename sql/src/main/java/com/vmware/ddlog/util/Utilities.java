package com.vmware.ddlog.util;

import java.util.*;

public class Utilities {
    public static Set<String> makeSet(String... data) {
        return new HashSet<String>(Arrays.asList(data));
     }

     public static <T> List<T> concatenate(List<T>... lists) {
        ArrayList<T> result = new ArrayList<T>();
        for (List<T> l: lists)
            result.addAll(l);
        return result;
     }
}
