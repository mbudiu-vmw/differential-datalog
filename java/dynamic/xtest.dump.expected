From 0 Insert O{10}
